### ISSUE TITLE
Question about using loop in loss function

### ISSUE THREAD

**Issue by kuanchiun** [FluxML/Flux.jl#2280]
Hello!

I wonder if it is possible to use a loop in the loss function?
Here is my current code
```
function eval_loss(x, encoder, neighbors)
    # Predict x̂
    x̂ = predict(x, encoder)
    
    # Reshape x and x̂ and calculate x̂ - x
    x_vec = reshape(x, 420, 2, 936)
    x̂_vec = reshape(x̂, 420, 2, 936)
    diff_vec = x̂_vec - x_vec
    
    # Count cosine similarity loss
    cos_loss = 0
    for i in 1:size(x_vec)[3]
        for j in 1:size(x_vec)[1]
            cos_sim = 0
            for k in 1:20
                cos = cosine_similarity(x̂_vec[j, :, i], x_vec[neighbors[j, k, i], :, i] - x_vec[j, :, i])
                if cos > cos_sim
                    cos_sim = cos
                end
            end
            cos_loss += (1 - cos_sim)
        end
    end
    
    return cos_loss
end
```

The step of the code is
1. predict x̂ from x
2. Reshape x and x̂ to the size (m, 2, n) what I need, and calculate (x̂ - x)
3. Consider each (1, 2, 1) as a point, and find it neighbor through another 3-dim array by its index in x
4. calculate (neighbor - x) for each neighbor, and measure the maximum cosine similarity with (x̂ - x)
5. add (1 - maximum cosine similarity) in cos_loss

Do you think that this loss function is possible for evaluate loss and do gradient descent?
Or is there any other strategy to reconstruct it without for loop?

Many thanks!
Kuan Chiun

**Comment by CarloLucibello** [comment]
Hi, 
[discourse](https://discourse.julialang.org/) is a better place for usage questions...
Zygote can differentiate through loops. It's quite inefficient with indexing though. 
Maybe you can rewrite you code using NNlib.gather / NNlib.scatter operations. 
Also https://github.com/JuliaStats/Distances.jl could help with computing all distances at once.
